sdk/containers/azcontainerregistry/authentication_policy.go

```diff
@@ -32,93 +29,129 @@ const (
 type authenticationPolicyOptions struct {
 }
 
+// authenticationPolicy is a policy to do the challenge-based authentication for container registry service. The authorization flow is as follows:
+// Step 1: GET /api/v1/acr/repositories
+// Return Header: 401: www-authenticate header - Bearer realm="{url}",service="{serviceName}",scope="{scope}",error="invalid_token"
+// Step 2: Retrieve the serviceName, scope from the WWW-Authenticate header.
+// Step 3: POST /api/oauth2/exchange
+// Request Body : { service, scope, grant-type, aadToken with ARM scope }
+// Response Body: { refreshToken }
+// Step 4: POST /api/oauth2/token
+// Request Body: { refreshToken, scope, grant-type }
+// Response Body: { accessToken }
+// Step 5: GET /api/v1/acr/repositories
+// Request Header: { Bearer acrTokenAccess }
+// Each registry service shares one refresh token, it will be cached in refreshTokenCache until expire time.
+// Since the scope will be different for different API/repository/artifact, accessTokenCache will only work when continuously calling same API.
 type authenticationPolicy struct {
-	mainResource *temporal.Resource[azcore.AccessToken, acquiringResourceState]
-	cred         azcore.TokenCredential
-	aadScopes    []string
-	acrScope     string
-	acrService   string
-	authClient   *authenticationClient
+	refreshTokenCache *temporal.Resource[azcore.AccessToken, acquiringResourceState]
+	accessTokenCache  string
+	cred              azcore.TokenCredential
+	aadScopes         []string
+	authClient        *authenticationClient
 }
 
 func newAuthenticationPolicy(cred azcore.TokenCredential, scopes []string, authClient *authenticationClient, opts *authenticationPolicyOptions) *authenticationPolicy {
 	return &authenticationPolicy{
-		cred:         cred,
-		aadScopes:    scopes,
-		authClient:   authClient,
-		mainResource: temporal.NewResource(acquire),
+		cred:              cred,
+		aadScopes:         scopes,
+		authClient:        authClient,
+		refreshTokenCache: temporal.NewResource(acquireRefreshToken),
 	}
 }
 
 func (p *authenticationPolicy) Do(req *policy.Request) (*http.Response, error) {
-	// send a copy of the original request without body content
-	challengeReq, err := p.getChallengeRequest(*req)
-	if err != nil {
-		return nil, err
-	}
-	resp, err := challengeReq.Next()
-	if err != nil {
-		return nil, err
+	var resp *http.Response
+	var err error
+	if req.Raw().Header.Get(headerAuthorization) != "" {
+		// retry request could do the request with existed token directly
+		resp, err = req.Next()
+		if err != nil {
```

Comment:
Nit: you could move this and the other `err` checks outside the main `if` clause to consolidate them.